pubsub/fakepubsub/fake.go

```diff
@@ -0,0 +1,209 @@
+// Copyright 2018 The Go Cloud Authors
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//     https://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+// Package fakepubsub provides an in-memory fake pubsub implementation.
+// This should not be used for production: it is intended for local
+// development.
+//
+// fakepubsub does not support any types for As.
+package fakepubsub
+
+import (
+	"context"
+	"errors"
+	"sync"
+	"time"
+
+	"github.com/google/go-cloud/pubsub/driver"
+)
+
+type topic struct {
+	mu        sync.Mutex
+	subs      []*subscription
+	nextAckID int
+	closed    bool
+}
+
+// CreateTopic creates a new topic.
+// Create subscribers for the topic before publishing.
+func CreateTopic() driver.Topic {
+	return &topic{}
+}
+
+// SendBatch sends the messages to all the subscriptions for this topic. It is an
+// error if the topic is closed or has no subscriptions.
+func (t *topic) SendBatch(ctx context.Context, ms []*driver.Message) error {
```

Comment:
I made it an error because, at least on GCP, it is _not_ an error, and your message just disappears forever. In my experience, this is a confusing failure mode. Since this package is intended for testing, I thought it would help people to surface the problem clearly.

Reply:
This is an interesting one, because it's a choice between "better" behavior and consistency across implementations. E.g., what if the application actually does want to send messages when there are no subscribers? Maybe the user is running locally, and the app just sends messages; in the real system, some other application is receiving them.

When you say "intended for testing", what kind of testing are you thinking of? Integration tests? I think we also expect users to run their application locally using the "local" versions of our components, so think about that case as well.

Reply:
This seems like a good case for a sentinel error, so apps can check for this and ignore it if they want to.

Reply:
I'm assuming any sort of local execution is done for development and/or testing, not production use. This is "Go Cloud," after all. The way I like to use fakes is (a) for unit tests that get at system behavior that I can't control against the real system, like how errors are handled (assuming the fake has additional mock-like capabilities like error injection); and (b) to run my integration test suite through in `-short` mode, so I exercise my full integration test suite without hitting a real backend (see https://code.googlesource.com/gocloud/+/master/logging/logging_test.go for an example of that).

Back to the original question: I see your point, and I'll remove the error return.